app/src/org/commcare/logic/AndroidFormController.java

```diff
@@ -63,4 +65,9 @@ public void returnToStoredIndex() {
         this.formIndexToReturnTo = null;
     }
 
+    public void notifyPermission(String permission, boolean permissionGranted) {
+        if (mWidgetFactory != null) {
+            mWidgetFactory.notifyPermission(permission, permissionGranted);
```

Comment:
Passing the `notifyPermission()` call down through the `widgetFactory` doesn't seem right to me; a factory class shouldn't be used for something stateful like that. I think what I would do instead is the following:

In `FormEntryActivity`, there's a method called `getPendingWidget()` that searches the the current view's widgets for one that has registered a pending callout with the form controller. If you refactor the implementation of that method to delegate most of its work to be done in `AndroidFormController`, then you can reuse it here to get the widget that should be notified.